knowhere/index/vector_index/IndexDiskANNConfig.cpp

```diff
@@ -0,0 +1,201 @@
+#include <optional>
+#include <string>
+
+#include "index/vector_index/ConfAdapter.h"
+
+namespace knowhere {
+namespace {
+static constexpr const char* kMaxDegree = "max_degree";
+static constexpr const char* kSearchListSize = "search_list_size";
+static constexpr const char* kSearchDramBudgetGb = "search_dram_budget_gb";
+static constexpr const char* kBuildDramBudgetGb = "build_dram_budget_gb";
+static constexpr const char* kNumThreads = "num_threads";
+static constexpr const char* kPqDiskBytes = "pq_disk_bytes";
+
+static constexpr const char* kNumNodesToCache = "num_nodes_to_cache";
+static constexpr const char* kWarmUp = "warm_up";
+static constexpr const char* kUseBfsCache = "use_bfs_cache";
+
+static constexpr const char* kK = "k";
+static constexpr const char* kBeamwidth = "beamwidth";
+static constexpr const char* kCandidatesLimit = "candidates_limit";
+
+static constexpr const char* kRadius = "radius";
+static constexpr const char* kMinK = "min_k";
+static constexpr const char* kMaxK = "max_k";
+
+static constexpr const char* kDiskANNBuildConfig = "diskANN_build_config";
+static constexpr const char* kDiskANNPrepareConfig = "diskANN_prepare_config";
+static constexpr const char* kDiskANNQueryConfig = "diskANN_query_config";
+static constexpr const char* kDiskANNQueryByRangeConfig = "diskANN_query_by_range_config";
+
+/**
+ * @brief Check the numeric param's existence and type, and allocate it to the config.
+ *
+ * @tparam T can only be float or int.
+ * @param min_o the min value of the value range, should be std::nullopt if we want to use numeric_limits::min().
+ * @param max_o the max value of the value range, should be std::nullopt if we want to use numeric_limits::max().
+ * @throw KnowhereException if any error.
+ */
+template <typename T>
+void
+CheckNumericPararmAndSet(const Config& config, const std::string& key, std::optional<T> min_o, std::optional<T> max_o,
+                         T& to_be_set) {
+    static_assert(std::is_integral_v<T> || std::is_floating_point_v<T>, "CheckAndSet only accept int and float type");
+
+    T min = min_o.has_value() ? min_o.value() : std::numeric_limits<T>::min();
+    T max = max_o.has_value() ? max_o.value() : std::numeric_limits<T>::max();
+    if (std::is_same_v<T, float>) {
+        CheckFloatRange(config, key, min, max);
+    } else {
+        CheckIntegerRange(config, key, min, max);
+    }
+    config.at(key).get_to(to_be_set);
+}
+
+/**
+ * @brief Check the numeric param's existence and type, and allocate it to the config.
+ *
+ */
+void
+CheckBoolParmAndSet(const Config& config, const std::string& key, bool& to_be_set) {
+    if (!config.contains(key)) {
+        KNOWHERE_THROW_FORMAT("Param '%s' not exist", key.data());
+    }
+    if (!config[key].is_boolean()) {
+        KNOWHERE_THROW_FORMAT("Param '%s' should be a bool", key.data());
+    }
+    config.at(key).get_to(to_be_set);
+}
+}  // namespace
+
+void
+to_json(Config& config, const DiskANNBuildConfig& build_conf) {
```

Comment:
why not define this API like following, it's easy to use.
Config
to_json(const DiskANNBuildConfig& build_conf)

Reply:
This is not the API we define. The `to_json` and `from_json` are default signatures `nlohmann_json` use to serialize/deserialize object.